File: packages/anyq/test/errors.test.ts
import { describe, expect, test } from 'bun:test';
import {
  AnyonceQueueError,
  FingerprintMismatchError,
  InFlightError,
  isFingerprintMismatchError,
  isInFlightError,
  QueueConfigurationError,
} from '../src/errors';

describe('typed queue errors', () => {
  test('REQ-Q-1: an in-flight error carries the lease, the delay and a retryable flag', () => {
    const err = new InFlightError(1_700_000_030_000, 4_500);
    expect(err).toBeInstanceOf(AnyonceQueueError);
    expect(err.name).toBe('InFlightError');
    expect(err.code).toBe('in-flight');
    expect(err.retryable).toBe(true);
    expect(err.leaseUntil).toBe(1_700_000_030_000);
    expect(err.delayMs).toBe(4_500);
    expect(err.translated).toBe(false);
    expect(err.message).not.toContain('1_700_000_030_000');
  });

  test('REQ-Q-1: a mismatch error carries the record and is not retryable', () => {
    const record = {
      scope: 'orders/workers',
      key: 'm-1',
      fingerprint: 'aa',
      state: 'completed' as const,
      fence: 1,
      leaseUntil: 0,
      createdAt: 0,
      expiresAt: 1,
    };
    const err = new FingerprintMismatchError(record);
    expect(err.name).toBe('FingerprintMismatchError');
    expect(err.code).toBe('fingerprint-mismatch');
    expect(err.retryable).toBe(false);
    expect(err.record).toBe(record);
    expect(err.message).not.toContain('m-1');
  });

  test('REQ-Q-1: the predicates match a foreign copy of the error by code', () => {
    const foreign = Object.assign(new Error('other copy'), {
      code: 'in-flight',
      retryable: true,
      delayMs: 10,
    });
    expect(isInFlightError(foreign)).toBe(true);
    expect(isFingerprintMismatchError(foreign)).toBe(false);
    expect(isInFlightError(new QueueConfigurationError('no scope'))).toBe(false);
    expect(
      isFingerprintMismatchError(Object.assign(new Error('x'), { code: 'fingerprint-mismatch' })),
    ).toBe(true);
    expect(isInFlightError('not an error')).toBe(false);
  });
});

File: go/anyqmw/errors_test.go
package anyqmw_test

import (
	"errors"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/sns45/anyonce/go/anyonce"
	"github.com/sns45/anyonce/go/anyqmw"
)

func TestInFlightError(t *testing.T) {
	leaseUntil := time.Unix(1_700_000_030, 0)
	err := &anyqmw.InFlightError{LeaseUntil: leaseUntil, DelayMs: 4500}

	t.Run("REQ-Q-1: an in-flight error carries the lease and the delay and reaches its sentinel", func(t *testing.T) {
		if !errors.Is(err, anyqmw.ErrInFlight) {
			t.Fatal("errors.Is did not reach ErrInFlight")
		}
		if errors.Is(err, anyqmw.ErrFingerprintMismatch) {
			t.Fatal("errors.Is reached the wrong sentinel")
		}
		var target *anyqmw.InFlightError
		if !errors.As(fmt.Errorf("wrapped: %w", err), &target) || target != err {
			t.Fatal("errors.As did not reach the concrete type through a wrapper")
		}
		if !target.LeaseUntil.Equal(leaseUntil) || target.DelayMs != 4500 {
			t.Fatalf("the error lost its lease data: %+v", target)
		}
	})

	t.Run("REQ-Q-1: an in-flight error is retryable and names no key, scope or payload", func(t *testing.T) {
		if !err.Retryable() {
			t.Fatal("an in-flight duplicate should be retryable")
		}
		message := err.Error()
		for _, secret := range []string{"m-1", "orders", "1700000030", "4500"} {
			if strings.Contains(message, secret) {
				t.Fatalf("the error message leaks %q: %s", secret, message)
			}
		}
	})

	t.Run("REQ-Q-8: an in-flight error records that a strategy translated it", func(t *testing.T) {
		fresh := &anyqmw.InFlightError{LeaseUntil: leaseUntil, DelayMs: 1}
		if fresh.Translated() {
			t.Fatal("a fresh error is already marked translated")
		}
		fresh.MarkTranslated()
		if !fresh.Translated() {
			t.Fatal("MarkTranslated did not stick")
		}
	})
}

func TestMismatchError(t *testing.T) {
	record := &anyonce.Record{Scope: "orders/workers", Key: "m-1", Fingerprint: "aa", State: anyonce.StateCompleted, Fence: 1}
	err := &anyqmw.MismatchError{Record: record}

	t.Run("REQ-Q-4: a mismatch error carries the record and reaches its sentinel", func(t *testing.T) {
		if !errors.Is(err, anyqmw.ErrFingerprintMismatch) {
			t.Fatal("errors.Is did not reach ErrFingerprintMismatch")
		}
		if errors.Is(err, anyqmw.ErrInFlight) {
			t.Fatal("errors.Is reached the wrong sentinel")
		}
		var target *anyqmw.MismatchError
		if !errors.As(fmt.Errorf("wrapped: %w", err), &target) || target.Record != record {
			t.Fatal("errors.As did not reach the concrete type with its record")
		}
	})

	t.Run("REQ-Q-4: a mismatch error is not retryable and names no key, scope or fingerprint", func(t *testing.T) {
		if err.Retryable() {
			t.Fatal("a payload mismatch should not be retryable")
		}
		message := err.Error()
		for _, secret := range []string{"m-1", "orders", "aa"} {
			if strings.Contains(message, secret) {
				t.Fatalf("the error message leaks %q: %s", secret, message)
			}
		}
	})
}

func TestErrConfiguration(t *testing.T) {
	t.Run("REQ-Q-1: a configuration failure reaches ErrConfiguration through a wrapper", func(t *testing.T) {
		err := fmt.Errorf("%w: set Options.Scope", anyqmw.ErrConfiguration)
		if !errors.Is(err, anyqmw.ErrConfiguration) {
			t.Fatal("errors.Is did not reach ErrConfiguration")
		}
		if errors.Is(err, anyqmw.ErrInFlight) || errors.Is(err, anyqmw.ErrFingerprintMismatch) {
			t.Fatal("errors.Is reached the wrong sentinel")
		}
	})
}
